pghoard/transfer.py

```diff
@@ -324,9 +457,20 @@ def handle_upload(self, site, key, file_to_transfer: UploadEvent):
                 metadata = file_to_transfer.metadata.copy()
                 if file_to_transfer.file_size:
                     metadata["Content-Length"] = str(file_to_transfer.file_size)
+                self.upload_tracker.track_upload_event(
+                    file_key=key,
+                    file_type=file_to_transfer.file_type,
+                    file_size=file_to_transfer.file_size,
+                )
+                upload_progress_fn = partial(self.upload_tracker.increment, file_key=key)
                 storage.store_file_object(
-                    key, f, metadata=metadata, upload_progress_fn=file_to_transfer.incremental_progress_callback
+                    key,
+                    f,
+                    metadata=metadata,
+                    upload_progress_fn=lambda n_bytes: upload_progress_fn(total_bytes_uploaded=n_bytes),
                 )
+                # make sure we untrack it after it is done
+                self.upload_tracker.untrack_upload_event(key)
```

Comment:
Need try/finally or make upload tracker support context manager and use `with` here. Now failed uploads are not unregistered.

Reply:
untracked it when handling exception

Reply:
That still doesn't look especially safe. Exception may be raised before the tracking starts or after it has ended so we may call untrack for a transfer which isn't tracked. I didn't check whether that actually causes problems or not but isn't especially clean anyway. The exception block could also end up raising before it gets as far as untracking the transfer. Simple context manager so that you could just say `with self.upload_tracker.track_upload(key): ...` would be clearly the cleanest code.

Reply:
it shouldn't cause any problem, if it doesn't exist on the dict of tracked events it should just skip it and if the TransferAgent tries to "re-track" it, then it'll just replace the previous one. I thought it was a bit "cleaner"  and simple to just add it to the except block since we "re-upload" the event there. But can try adding a context manager